### PR TITLE
ScaffoldGeometry plumbing.

### PR DESCRIPTION
**Pull request by amirh**
Adds a ScaffoldGeometry class and ValueNotifier for it.
A scaffold's ScaffoldGeometry notifier is held in the _ScaffoldState,
and is passed to _ScaffoldScope.
New ScaffoldGemometry objects are built and published to the notifier
during the layout pass after computing the scaffold's layout.

### CODE REVIEW

**Comment by amirh** [comment]
@Hixie @DaveShuckerow @HansMuller here's a quick suggestion of how to plumb the ScaffoldGeometry change notifications.

Not tests yet, as I'm first seeking your feedback on the way it's plumbed.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 1]
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
   statusBar,
 }
 
+/// Geometry information for scaffold components.
+///
+/// To get a [ValueNotifier] for the scaffold gemoetry call
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: (typo in geometry)

File: packages/flutter/lib/src/material/scaffold.dart [hunk 2]
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
...
+  /// [Scaffold.bottomNavigationBar].
+  /// Null when there is no bottom navigation bar..
+  final Rect bottomNavigationBar;
+
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.floatingActionButton].
+  /// Null when there is no bottom navigation bar..
+  final Rect floatingActionButton;
+
+  const ScaffoldGeometry({
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: nit: constructors first (when first reading this i assumed there was an implicit constructor, but when i got to this constructor i realised no, actually, there's no implicit constructor)

File: packages/flutter/lib/src/material/scaffold.dart [hunk 3]
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
...
+/// Geometry information for scaffold components.
+///
+/// To get a [ValueNotifier] for the scaffold gemoetry call
+/// [Scaffold.geometryOf].
+@immutable
+class ScaffoldGeometry {
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.bottomNavigationBar].
+  /// Null when there is no bottom navigation bar..
+  final Rect bottomNavigationBar;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: This could probably just be a double giving the distance from some reference edge to the top of the navbar area. That would reduce the temptation to use this value to determine geometry that is best discovered by directly examining the navbar.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 4]
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
...
+
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.floatingActionButton].
+  /// Null when there is no bottom navigation bar..
+  final Rect floatingActionButton;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: similarly here i'd probably express this in terms of the offset from some reference point to the center of the fab, and a double giving the current fractional size of the fab. That would give more flexibility, e.g. in case the shape of the fab is not well described by a Rect (e.g. it's a T-shape, or a diamond).

File: packages/flutter/lib/src/material/scaffold.dart [hunk 5]
```diff
@@ -514,6 +548,35 @@ class Scaffold extends StatefulWidget {
     );
   }
 
+  /// Returns a [ValueNotifier] for the [ScaffoldGeometry] for the closest
+  /// [Scaffold] ancestor of the given context.
+  ///
+  /// Value notifications are sent during the layout pass.
+  static ValueNotifier<ScaffoldGeometry> geometryOf(BuildContext context) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: to avoid anyone being tempted to set the value directly on this object, maybe make this return a ValueListenable instead of a ValueNotifier. (ValueNotifier implements ValueListenable so no code need change.)

File: packages/flutter/lib/src/material/scaffold.dart [hunk 6]
```diff
@@ -514,6 +548,35 @@ class Scaffold extends StatefulWidget {
     );
   }
 
+  /// Returns a [ValueNotifier] for the [ScaffoldGeometry] for the closest
+  /// [Scaffold] ancestor of the given context.
+  ///
+  /// Value notifications are sent during the layout pass.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: docs should talk about how this object must be discarded any time the given context has its dependencies change, notably, for the context from a StatefulWidget, whenever the corresponding State's didChangeDepndencies is fired.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 7]
```diff
@@ -514,6 +548,35 @@ class Scaffold extends StatefulWidget {
...
+  static ValueNotifier<ScaffoldGeometry> geometryOf(BuildContext context) {
+    final _ScaffoldScope scaffoldScope = context.inheritFromWidgetOfExactType(_ScaffoldScope);
+    if (scaffoldScope == null)
+    throw new FlutterError(
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: nits: indent, braces

File: packages/flutter/lib/src/material/scaffold.dart [hunk 8]
```diff
@@ -514,6 +548,35 @@ class Scaffold extends StatefulWidget {
...
+      'context that is "under" the Scaffold. For an example of this, please see the '
+      'documentation for Scaffold.of():\n'
+      '  https://docs.flutter.io/flutter/material/Scaffold/of.html\n'
+      'A more efficient solution is to split your build function into several widgets. This '
+      'introduces a new context from which you can obtain the Scaffold. In this solution, '
+      'you would have an outer widget that creates the Scaffold populated by instances of '
+      'your new inner widgets, and then in these inner widgets you would use Scaffold.geometryOf().\n'
+      'A less elegant but more expedient solution is assign a GlobalKey to the Scaffold, '
+      'then use the key.currentState property to obtain the ScaffoldState rather than '
+      'using the Scaffold.geometryOf() function.\n'
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: this message doesn't make sense. You can't use a GlobalKey to get the geometry value notifier.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 9]
```diff
@@ -36,18 +36,42 @@ enum _ScaffoldSlot {
...
 
+/// Geometry information for scaffold components.
+///
+/// To get a [ValueNotifier] for the scaffold gemoetry call
+/// [Scaffold.geometryOf].
+@immutable
+class ScaffoldGeometry {
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.bottomNavigationBar].
+  /// Null when there is no bottom navigation bar..
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: nit: these docs need work ("Null" should be lowercase, since Null is the type and null is the value; double period at the end; blank line should be between the two sentences, needs more detail)

**Comment by Hixie** [comment]
I recommended some trivial tweaks, but the general approach seems solid to me.

One thing I would look into is how the fab shrink/grow logic is implemented. Somehow that needs to be exposed in this geometry object; it's possible we implement that efficiently enough that it's entirely done as an animation/paint loop, no layout, in which case we presumably want to somehow get the relevant Animation involved here.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 2 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart (hunk 6 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: Added some details to the doc

File: packages/flutter/lib/src/material/scaffold.dart (hunk 8 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: oops, clumsy copy-pasting

File: packages/flutter/lib/src/material/scaffold.dart (hunk 7 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart (hunk 9 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: change the nits, though I'm not sure which more details are missing here.
(though we might be changing this to not be Rect at all, so maybe not relevant)

File: packages/flutter/lib/src/material/scaffold.dart (hunk 1 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/scaffold.dart]: In the fullness of time, this doc should explain what "geometry" means in this context, why you'd want to listen to this notifier, and when you'd want to handle notifications.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 3 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart (hunk 4 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: This will be a fraction of what? of some fixed fab size? if we want to reduce this to a single double maybe we can use the size of one of the edges, or the diagonal, though I'm not sure this the best thing to do here...

Your point for reducing the temptation to use the bottomNavigationBarRect makes sense when thinking of the BAB use case, where we want to at least look at the BAB's size.

For the FAB dimensions though, there is no such temptation, in the BAB use case all you know about the FAB comes from ScaffoldGeometry.

I would leave it a Rect as it contains the full information of where the scaffold is laying out the FAB in a simple form. Also you could think of a case where the width/height proportions are changing, and a fractional size is not enough.

**Comment by amirh** [comment]
Good point regarding the FAB scaling logic, it is indeed implemented with a Transform that's skipping layout. I'm make it update geometry as well, though it makes its contract a bit more complex (we can no longer say it's updated as part of the layout pass)

**Comment by amirh** [comment]
PTAL, added tests, and added geometry notifications from the FAB transition animation.

TBD in following CLs - add the 'notchMaker' closure we discussed to ScaffoldGeometry

File: packages/flutter/lib/src/material/scaffold.dart [hunk 10]
```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
...
 
+/// Geometry information for scaffold components.
+///
+/// To get a [ValueNotifier] for the scaffold geometry call
+/// [Scaffold.geometryOf].
+@immutable
+class ScaffoldGeometry {
+  const ScaffoldGeometry({
+    this.bottomNavigationBarTop,
+    this.floatingActionButton,
```
Comment (DaveShuckerow) [inline packages/flutter/lib/src/material/scaffold.dart]: I think we're going to need two different types of ScaffoldGeometry.  The fab placement logic I'm working on depends on the scaffold's geometry for figuring out what offset to give the fab, so the position rect won't apply.

I think the easiest thing to do will be to pass the fab positioner an incomplete scaffold geometry, then we can finish populating it with the fab layout after it's computed.  

### TLDR
So long as it's possible to build ScaffoldGeometry without data about the fab's layout and without passing the incomplete object into the value notifier, this shouldn't be a problem.

Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: Will we be able to populate the floatingActionButton with it's position before paint time?

As long as by paint time we populated the fab information in the geometry object it should be ok.
In current CL we already populate it with incomplete data from the FabTransition before the first layout.

Comment (DaveShuckerow) [reply packages/flutter/lib/src/material/scaffold.dart]: Yes.  The way it works in #14368 currently is:

1. Scaffold passes FabPositioner and FabMotionAnimator to the Layout delegate
2. Layout delegate computes layout of everything but the fab and creates a ScaffoldGeometry
3. FabPositioner uses the ScaffoldGeometry to determine the Fab's Offset
4. Layout delegate computes the Fab's layout based on the requested offset (interpolated by animation when necessary)

The output of this can be a ScaffoldGeometry that includes the Fab's layout rect, and that will be accessible as part of the Layout delegate's layout phase, before painting.

Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: great

Comment (Hixie) [reply packages/flutter/lib/src/material/scaffold.dart]: it seems weird to me to use the same struct for computing the fab position and for obtaining the fab position

File: packages/flutter/lib/src/material/scaffold.dart [hunk 11]
```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
...
+  ///
+  /// When there is no [Scaffold.bottomNavigationBar] set, this will be null.
+  final double bottomNavigationBarTop;
+
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.floatingActionButton].
+  ///
+  /// This can have any value when there is no [Scaffold.floatingActionButton]
+  /// set.
+  final Rect floatingActionButton;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: is a Rect really the best way to describe this? Consider a FAB whose shape is a downwards-pointing triangle.

I'd probably just go with an offset to the center, and a radius?

Anyway, a Rect is fine if you think it's better than that.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 12]
```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
...
+
+  /// The amount by which the [Scaffold.floatingActionButton] is scaled.
+  final double floatingActionButtonScale;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: This seems redundant with the Rect?

Comment (Hixie) [reply packages/flutter/lib/src/material/scaffold.dart]: At a minimum the docs should explain in detail how the scale and the position of the fab interact.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 11 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: we should probably call this floatingActionButtonPosition or something

File: packages/flutter/lib/src/material/scaffold.dart [hunk 13]
```diff
@@ -36,18 +36,63 @@ enum _ScaffoldSlot {
...
+
+  static ScaffoldGeometry _update({
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: this should just use the copyWith pattern

File: packages/flutter/lib/src/material/scaffold.dart [hunk 14]
```diff
@@ -284,6 +347,24 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
...
+      return;
+    }
+    if (_currentAnimation.status != AnimationStatus.dismissed) {
+      _updateGeometryScale(_currentAnimation.value);
+      return;
+    }
+  }
+
+  void _updateGeometryScale(double scale) {
+    widget.geometryNotifier.value = ScaffoldGeometry._update(
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: this is going to cause the notifier to notify multiple times per layout, which seems suboptimal.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 15]
```diff
@@ -514,6 +595,50 @@ class Scaffold extends StatefulWidget {
...
+  /// using this listenable, a change of the [Scaffold] ancestor will
+  /// trigger a [State.didChangeDependencies].
+  ///
+  /// A typical pattern for listening to the scaffold geometry would be to
+  /// call [Scaffold.geometryOf] in [State.didChangeDependencies], compare the
+  /// return value with the previous listenable, if it has changed, unregister
+  /// the listener, and register a listener to the new [ScaffoldGeometry]
+  /// listenable.
+  // TODO(amirh): return a ValueListenable subclass that asserts it's paint time
+  // when .value is accessed.
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: we should just do that now, otherwise we'll never be able to do it because people will depend on us not doing it

File: packages/flutter/lib/src/material/scaffold.dart [hunk 16]
```diff
@@ -798,6 +925,12 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
 
   // INTERNALS
 
+  @override
+  void initState() {
+    super.initState();
+    _geometryNotifier = new ValueNotifier<ScaffoldGeometry>(null);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: dispose of this in the dispose method

File: packages/flutter/lib/src/material/scaffold.dart [hunk 17]
```diff
@@ -212,15 +270,20 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
...
     if (widget.child != null)
       _currentController.value = 1.0;
+    else
+      _updateGeometryScale(_previousAnimation.value);
   }
 
   @override
   void dispose() {
+    _previousAnimation.removeListener(_onProgressChanged);
+    _currentAnimation.removeListener(_onProgressChanged);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: you don't need to remove these, the dispose methods below will do it for you

File: packages/flutter/lib/src/material/scaffold.dart (hunk 14 above)
Comment (Hixie) [reply packages/flutter/lib/src/material/scaffold.dart]: we discussed this and couldn't find a workaround.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 11 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: Following f2f we keep it a rect, as this keeps the notch computation logic similar to the path computation logic for the ShapeBorder (both take a Rect as their input)

File: packages/flutter/lib/src/material/scaffold.dart (hunk 12 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: Added docs

File: packages/flutter/lib/src/material/scaffold.dart (hunk 13 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: renamed to _copyWith, is that what you meant?

File: packages/flutter/lib/src/material/scaffold.dart (hunk 16 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart (hunk 17 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart (hunk 15 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

**Comment by amirh** [comment]
PTAL

File: packages/flutter/lib/src/material/scaffold.dart [hunk 18]
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
...
+/// Geometry information for scaffold components.
+///
+/// To get a [ValueNotifier] for the scaffold geometry call
+/// [Scaffold.geometryOf].
+@immutable
+class ScaffoldGeometry {
+  const ScaffoldGeometry({
+    this.bottomNavigationBarTop,
+    this.floatingActionButtonPosition,
+    this.floatingActionButtonScale = 1.0,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: :

File: packages/flutter/lib/src/material/scaffold.dart [hunk 19]
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
...
+  ///
+  /// When there is no [Scaffold.bottomNavigationBar] set, this will be null.
+  final double bottomNavigationBarTop;
+
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.floatingActionButton].
+  ///
+  /// The floating action button might be scaled inside this rectangle, to get
+  /// the bounding rectangle in which the floating action is painted scale this
+  /// value by [floatingActionButtonScale].
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: It's not clear what "scale" means here. Maybe show same `## Sample code` for how to do that?

File: packages/flutter/lib/src/material/scaffold.dart [hunk 20]
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
...
+  }
+}
+
+class _ScaffoldGeometryNotifier extends ValueNotifier<ScaffoldGeometry> {
+  _ScaffoldGeometryNotifier(ScaffoldGeometry geometry)
+    : super(geometry);
+
+  @override
+  ScaffoldGeometry get value{
+    if (!RendererBinding.instance.pipelineOwner.debugDoingPaint)
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: that might not be the right PipelineOwner.

You want the owner for the RenderObject that's actually doing the scaffold layout.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 21]
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
...
+      throw new FlutterError(
+        'Scaffold.geometryOf() value was accessed not during the paint phase.'
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: we should phrase this as something like "Scaffold.geometryOf() value must only be accessed during the paint phase.", otherwise the logical reaction to that error message is "I know, so what".

Comment (Hixie) [reply packages/flutter/lib/src/material/scaffold.dart]: Maybe even make this a multiline message:

```
Scaffold.geometryOf() value obtained outside paint phase.
The ScaffoldGeometry is only available during the paint phase, because its value is computed during
the animation and layout phases prior to painting. If you are trying to obtain the value during the
build or layout phases, consider moving your logic to the paint phase, since the value is not yet
completely available during build or layout. If you are trying to obtain the value between frames,
consider caching the value during paint.
```

...or something

File: packages/flutter/lib/src/material/scaffold.dart [hunk 22]
```diff
@@ -212,11 +294,14 @@ class _FloatingActionButtonTransitionState extends State<_FloatingActionButtonTr
       parent: _currentController,
       curve: Curves.easeIn
     );
+    _currentAnimation.addListener(_onProgressChanged);
 
     // If we start out with a child, have the child appear fully visible instead
     // of animating in.
     if (widget.child != null)
       _currentController.value = 1.0;
+    else
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: there's a comment above explaining the if statement but its explanation is now incomplete

File: packages/flutter/lib/src/material/scaffold.dart [hunk 23]
```diff
@@ -548,6 +693,8 @@ class Scaffold extends StatefulWidget {
 /// the current [BuildContext] using [Scaffold.of].
 class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin {
 
+  _ScaffoldGeometryNotifier _geometryNotifier;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: we should put this in the section of this class that uses it, i.e. below the `// INTERNALS` comment above initState

File: packages/flutter/lib/src/material/scaffold.dart (hunk 18 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart (hunk 19 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done.
@Hixie looking at it again I don't like the Position suffix we added to floatingActionButton, as it sounds like a point and not a rectangle. What do you think about just calling it floatingActionButton (geometry.floatingActionButton sounds pretty reasonable I believe).
If not, maybe floatingActionButtonRect ?

File: packages/flutter/lib/src/material/scaffold.dart [hunk 24]
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
...
+  final double bottomNavigationBarTop;
+
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.floatingActionButton].
+  ///
+  /// The floating action button might be scaled inside this rectangle, to get
+  /// the bounding rectangle in which the floating action is painted scale this
+  /// value by [floatingActionButtonScale].
+  ///
+  /// This can have any value when there is no [Scaffold.floatingActionButton]
```
[the remark this answers is not in the document]
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: For the BAB needs it is not a problem: when there is no FAB the scale is 0, so the BAB won't create a notch.

Setting this to null when there's no FAB makes the dance between the transition animation ticks and the layout code even more complex (when you set the floatingActionButton to null, there is still a FAB until the transition animation is complete). Which I rather avoid at this stage (we expect some more churn to this code with Dave's work, once it stabilize we can see if it's reasonable).

I added to the doc comment that when there is no FAB the scale will be 0.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 25]
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
...
+  /// set.
+  final Rect floatingActionButtonPosition;
+
+  /// The amount by which the [Scaffold.floatingActionButton] is scaled.
+  ///
+  /// To et the bounding rectangle in shich the floating action button is
```
[the remark this answers is not in the document]
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart [hunk 26]
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
...
+  /// painted scale [floatingActionPosition] by this proportion.
```
[the remark this answers is not in the document]
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart [hunk 27]
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
...
+  /// set.
+  final Rect floatingActionButtonPosition;
+
+  /// The amount by which the [Scaffold.floatingActionButton] is scaled.
+  ///
+  /// To et the bounding rectangle in shich the floating action button is
+  /// painted scale [floatingActionPosition] by this proportion. 
+  final double floatingActionButtonScale;
+
+  static ScaffoldGeometry _copyWith({
```
[the remark this answers is not in the document]
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: @Hixie  I'm sorry but I do not know what do you mean by the "copywith pattern", can you be more explicit? do you mean you want the method non static? it won't work in this case as previousGeometry might be null.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 28]
```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
...
+    );
+  }
+}
+
+class _ScaffoldGeometryNotifier extends ValueNotifier<ScaffoldGeometry> {
+  _ScaffoldGeometryNotifier(ScaffoldGeometry geometry)
+    : super(geometry);
+
+  @override
+  ScaffoldGeometry get value{
```
[the remark this answers is not in the document]
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart (hunk 20 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done.
I'm not sure how expensive is the findRenderObject, should/can I cache the pipeline owner once I find it?

File: packages/flutter/lib/src/material/scaffold.dart (hunk 21 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: I sub-sampled parts of your message :)

File: packages/flutter/lib/src/material/scaffold.dart (hunk 22 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

**Comment by amirh** [comment]
PTAL

File: packages/flutter/lib/src/material/scaffold.dart (hunk 27 above)
[the remark this answers is not in the document]
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: I moved the partial update logic into _ScaffoldGeometryNotifier, this way the partial value is not accessible even from the scaffold code.

Still curious to know what do you mean by the "copywith pattern"

File: packages/flutter/lib/src/material/scaffold.dart (hunk 20 above)
Comment (Hixie) [reply packages/flutter/lib/src/material/scaffold.dart]: Just put all this code in an assert.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 27 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/material/scaffold.dart]: If you look around for `copyWith` methods you'll see what I mean. Lots of our classes have a method named `copyWith` that allows the class to be copied with edits.

To solve the "null" problem, just initialize the ValueNotifier with a non-null value.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 19 above)
Comment (Hixie) [reply packages/flutter/lib/src/material/scaffold.dart]: A "floatingActionButton" property should hold a floating action button.
I agree that "position" suggests an Offset.

This gives you the area that the FAB would cover if it was not scaled, right?
Maybe "floatingActionButtonArea"?

File: packages/flutter/lib/src/material/scaffold.dart (hunk 24 above)
[the remark this answers is not in the document]
Comment (Hixie) [reply packages/flutter/lib/src/material/scaffold.dart]: I guarantee you that eventually someone will write code that depends on exactly what value you export here when the scale is 0. If we're going to expose a non-null value there in that case, I think we should at a minimum document exactly what the value will be, even if it isn't very useful. The reality is that eventually (maybe in a few years) we'll be forced to be compatible with whatever value we expose here even if we want to change it.

File: packages/flutter/lib/src/material/scaffold.dart (hunk 23 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart (hunk 20 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done

**Comment by amirh** [comment]
PTAL

File: packages/flutter/lib/src/material/scaffold.dart [hunk 29]
```diff
@@ -36,18 +37,112 @@ enum _ScaffoldSlot {
...
+
+  final BuildContext context;
+
+  @override
+  ScaffoldGeometry get value {
+    assert(() {
+      final RenderObject renderObject = context.findRenderObject();
+      if (renderObject == null || !renderObject.owner.debugDoingPaint)
+        throw new FlutterError(
+            'Scaffold.geometryOf() must only be accessed during the paint phase.\n\n'
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: only one `\n` after the first line (see also `FlutterError.message` documentation)

File: packages/flutter/lib/src/material/scaffold.dart [hunk 30]
```diff
@@ -36,18 +37,112 @@ enum _ScaffoldSlot {
...
+                'The ScaffoldGeometry is only available during the paint phase, because\n'
+                'its value is computed during the animation and layout phases prior to painting.'
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: inconsistent indentation

File: packages/flutter/lib/src/material/scaffold.dart [hunk 31]
```diff
@@ -36,18 +37,112 @@ enum _ScaffoldSlot {
...
+    return super.value;
+  }
+
+  void _updateWith({
+    double bottomNavigationBarTop,
+    Rect floatingActionButtonArea,
+    double floatingActionButtonScale,
+  }) {
+
+    final double newFloatingActionButtonScale = floatingActionButtonScale ??  super.value?.floatingActionButtonScale;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: nit: extra spaces here and below after `??`

File: packages/flutter/lib/src/material/scaffold.dart [hunk 32]
```diff
@@ -36,18 +37,112 @@ enum _ScaffoldSlot {
...
+    }());
+    return super.value;
+  }
+
+  void _updateWith({
+    double bottomNavigationBarTop,
+    Rect floatingActionButtonArea,
+    double floatingActionButtonScale,
+  }) {
+
```
Comment (Hixie) [inline packages/flutter/lib/src/material/scaffold.dart]: nit: extraneous blank line

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/material/scaffold.dart (hunk 29 above)
Comment (amirh) [reply packages/flutter/lib/src/material/scaffold.dart]: done
